### ISSUE TITLE
Trigger multiple pipelines for one commit

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/ods-core#61]
It is possible that one repo contains more than one `Jenkinsfile`. Now that we have a webhook proxy, it is relatively trivial to allow users to specify the locaction of each `Jenkinsfile`, and create one pipeline for each. This would help to deal e.g. with [multi module projects](https://spring.io/guides/gs/multi-module/).

The implementation would roughly look like this:

Allow to add a query param to the webhook, e.g. `?dirs=foo,bar`. In that case, the webhook proxy would create two pipelines, one pointing to `foo/Jenkinsfile` and the other pointing to `bar/Jenkinsfile`. Then it would trigger both pipelines.

Futher, each pipeline could check whether it really needs to run. It would do that by comparing if something in the folder changed compared to the last build of that folder. First, it would get the latest built commit sha via `oc -n foo-dev get -o template bc/be-main --template={{.spec.output.to.name}}`, and then it would check for changes via `git diff --exit-code <sha1>..<sha2> <folder>`.

@tjaeschke @rattermeyer @stitakis @clemensutschig Thoughts on this?

**Comment by rattermeyer** [comment]
Hmm. Not conviced yet. Wouldn't be another option to create a "Master" Jenkinsfile in a multi-module build and call the other Jenkinsfile, like proposed here: https://stackoverflow.com/questions/44766324/invoke-a-remote-jenkins-file-from-another-jenkinsfile.
You would not get a seperate pipeline for each module, but if you have a multi-module project, I would expect that the different modules are released and deployed together. In this case, I would like to have a single pipeline for the project and only optimize build times as you have described.

**Comment by michaelsauter** [comment]
@rattermeyer Nice find, did not know this is possible.

I am not convinced however that one would want to deploy the modules together. There are many who favour a monorepo over many repos, but that does not mean that you build a monolith (everything deployed at the same time) necessarily. Therfore I think it could make sense to have a different pipeline.

That said, I I'll think a bit more about your proposal. Also, we can wait until we have a few use cases and then decide on the best solution for this.

**Comment by rattermeyer** [comment]
Monorepo <> Multi-module project. If we talk about a monorepo, then your idea makes sense. But a monorepo should /can contain independent projects. But a multi-module build tends to have the same version for all components.

**Comment by michaelsauter** [comment]
As this came up in a slightly different discussion again, the solution has evolved slightly (@sino92 thanks for the hint!):

Instead of allowing to specify multiple jenkinsfiles in one webhook, we just add a feature to specify one jenkinsfile. Doing so:

* Allows users to put the Jenkinsfile somewhere else than in the root dir
* Allows users to create multiple webhooks, firing for potentially different events, to many Jenkinsfiles, each of which can determine if they want to run or not.

I think this is a great solution, giving a lot of flexibility and very straight-forward to implement.

**Comment by stitakis** [comment]
@michaelsauter idea sounds good to me but not sure if I understood it correctly... the proposed change would allow users to specify in a webhook the path and name of the Jenkinsfile, which ev. will be located not in root dir, to be triggered by Jenkins, right?

**Comment by michaelsauter** [comment]
@stitakis Yes, it would allow users to configure https://github.com/opendevstack/ods-core/blob/master/jenkins/webhook-proxy/pipeline.json#L31 by using a `GET` param in the webhook URL.
